Userland/Libraries/LibJS/Tests/syntax/function-hoisting.js

```diff
@@ -0,0 +1,352 @@
+
+            function hoistSecondOneTL() {
+                return "hoistThirdOneTL";
+            }
+
+            expect(hoistSecondOneTL()).toEqual("hoistThirdOneTL");
+        }
+
+        expect(hoistSecondOneTL()).toEqual("hoistSecondOneTL");
+    }
+
+    expect(hoistSecondOneTL()).toEqual("hoistSecondOneTL");
+}
+
+expect(hoistSecondOneTL()).toEqual("hoistSecondOneTL");
+
+test("Non-strict function does hoist", () => {
+    const noHoistLexFunc = false;
+    let canCallNonHoisted = 0;
+
+    expect(basicHoistFunc()).toEqual("basicHoistFunc");
+
+    function basicHoistFunc() {
+        return "basicHoistFunc";
+    }
+
+    expect(typeof noHoistLexFunc).toBe("boolean");
+    expect(typeof hoistInBlockFunc).toBe("undefined");
+
+    {
+        expect(noHoistLexFunc()).toEqual("noHoistLexFunc");
+        ++canCallNonHoisted;
+
+        expect(basicHoistFunc()).toEqual("basicHoistFuncInBlock");
+        ++canCallNonHoisted;
+
+        expect(hoistInBlockFunc()).toEqual("hoistInBlockFunc");
+
+        function hoistInBlockFunc() {
+            return "hoistInBlockFunc";
+        }
+
+        function noHoistLexFunc() {
+            return "noHoistLexFunc";
+        }
+
+        function basicHoistFunc() {
+            return "basicHoistFuncInBlock";
+        }
+    }
+
+    expect(canCallNonHoisted).toBe(2);
+
+    expect(hoistInBlockFunc()).toEqual("hoistInBlockFunc");
+
+    {
+        {
+            expect(nestedBlocksFunc()).toEqual("nestedBlocksFunc");
+
+            function nestedBlocksFunc() {
+                return "nestedBlocksFunc";
+            }
+        }
+        expect(nestedBlocksFunc()).toEqual("nestedBlocksFunc");
+    }
+    expect(nestedBlocksFunc()).toEqual("nestedBlocksFunc");
+    expect(hoistInBlockFunc()).toEqual("hoistInBlockFunc");
+
+    expect(typeof hoistSecondOneFunc).toBe("undefined");
+    {
+        expect(typeof hoistSecondOneFunc).toBe("undefined");
+
+        {
+            expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+
+            function hoistSecondOneFunc() {
+                return "hoistFirstOneFunc";
+            }
+
+            expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+
+            function hoistSecondOneFunc() {
+                return "hoistSecondOneFunc";
+            }
+
+            expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+
+            {
+                expect(hoistSecondOneFunc()).toEqual("hoistThirdOneFunc");
+
+                function hoistSecondOneFunc() {
+                    return "hoistThirdOneFunc";
+                }
+
+                expect(hoistSecondOneFunc()).toEqual("hoistThirdOneFunc");
+            }
+
+            expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+        }
+
+        expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+    }
+
+    expect(hoistSecondOneFunc()).toEqual("hoistSecondOneFunc");
+
+    expect(notBlockSFTL()).toBe("second");
+
+    function notBlockSFTL() {
```

Comment:
Please get rid of these acronyms everywhere, they don't exactly help with code readability :)